File: src/main/java/edu/cmu/ml/rtw/pra/graphs/RelationSet.java
package edu.cmu.ml.rtw.pra.graphs;

import java.io.BufferedReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;

import edu.cmu.ml.rtw.users.matt.util.Dictionary;
import edu.cmu.ml.rtw.users.matt.util.FileUtil;
import edu.cmu.ml.rtw.users.matt.util.IntTriple;
import edu.cmu.ml.rtw.users.matt.util.Pair;

public class RelationSet {

  // Fields dealing with the relations themselves.

  // The file containing the relation triples.
  private final String relationFile;
  // If not null, prepend this prefix to all relation strings in this set.
  private final String relationPrefix;
  // KB relations or surface relations?  The difference is only in the alias relationfinal .
  private final boolean isKb;

  // Fields specific to KB relations.

  // If this is a KB relation set, this file contains the mapping from entities to noun phrases.
  private final String aliasFile;
  // What should we call the alias relation?  Defaults to "alias".  Note that we specify this for
  // each set of _KB_ relations, not for each set of _surface_ relations.  If you want something
  // more complicated, like a different name for each (KB, surface) relation set pair, you'll have
  // to approximate it with relation prefixes.
  private final String aliasRelation;
  // Determines how we try to read the alias file.  We currently allow two values: "freebase" and
  // "nell".  The freebase format is a little complicated; the nell format is just a list of
  // (concept, noun phrase) pairs that we read into a map.  The "nell" format is generic enough
  // that it could be used for more than just NELL KBs, probably.
  private final String aliasFileFormat;

  // Fields specific to surface relations.

  // If this is a surface relation set, if this is true, we only add the alias edges that get
  // generated from the set, not the surface relations themselves.  This was for a random test that
  // I did.
  private final boolean aliasesOnly;

  // Fields for embeddings.

  // The file where we can find embeddings for the relations in this set.
  private final String embeddingsFile;
  // If we are embedding the edges, should we replace the original edges or just augment them?
  private final boolean keepOriginalEdges;

  private final FileUtil fileUtil;

  /**
   * Get the set of aliases specified by this KB relation set.
   */
  public Map<String, List<String>> getAliases() throws IOException {
    if (aliasFile == null) {
      return Maps.newHashMap();
    }
    return getAliasesFromReader(fileUtil.getBufferedReader(aliasFile));
  }

  @VisibleForTesting
  protected Map<String, List<String>> getAliasesFromReader(BufferedReader reader)
  throws IOException {
    if (aliasFileFormat.equals("freebase")) {
      return fileUtil.readMapListFromTsvReader(reader, 3, false, new FileUtil.LineFilter() {
        @Override
        public boolean filter(String[] fields) {
          if (fields.length != 4) return true;
          if (!fields[2].equals("/lang/en")) return true;
          return false;
        }
      });
    } else if (aliasFileFormat.equals("nell")) {
      return fileUtil.readInvertedMapListFromTsvReader(reader, 1000000);
    } else {
      throw new RuntimeException("Unrecognized alias file format");
    }
  }

  public int writeRelationEdgesToGraphFile(FileWriter intEdgeFile,
                                           Set<IntTriple> seenTriples,
                                           String prefixOverride,
                                           Set<String> seenNps,
                                           List<Pair<String, Map<String, List<String>>>> aliases,
                                           Dictionary nodeDict,
                                           Dictionary edgeDict) throws IOException {
    return writeRelationEdgesFromReader(fileUtil.getBufferedReader(relationFile),
                                        loadEmbeddings(),
                                        seenTriples,
                                        prefixOverride,
                                        seenNps,
                                        aliases,
                                        intEdgeFile,
                                        nodeDict,
                                        edgeDict);
  }

  @VisibleForTesting
  protected int writeRelationEdgesFromReader(BufferedReader reader,
                                             Map<String, List<String>> embeddings,
                                             Set<IntTriple> seenTriples,
                                             String prefixOverride,
                                             Set<String> seenNps,
                                             List<Pair<String, Map<String, List<String>>>> aliases,
                                             FileWriter writer,
                                             Dictionary nodeDict,
                                             Dictionary edgeDict) throws IOException {
    System.out.println("Adding edges from relation file " + relationFile);
    String prefix = "";
    if (prefixOverride != null) {
      prefix = prefixOverride;
    } else if (relationPrefix != null) {
      prefix = relationPrefix;
    }
    String line;
    int i = 0;
    int numEdges = 0;
    while ((line = reader.readLine()) != null) {
      fileUtil.logEvery(1000000, ++i);
      String[] fields = line.split("\t");
      String relation;
      String arg1;
      String arg2;
      // TODO(matt): Maybe make this relation file format a configurable field?
      if (isKb) {
        // KB relations are formated (S, O, V).
        arg1 = fields[0];
        arg2 = fields[1];
        relation = fields[2];
      } else {
        // And surface relations are formatted (S, V, O).
        arg1 = fields[0];
        relation = fields[1];
        arg2 = fields[2];
      }
      int ind1 = nodeDict.getIndex(arg1);
      int ind2 = nodeDict.getIndex(arg2);

      if (!isKb) {
        numEdges += addAliasEdges(arg1, ind1, seenNps, writer, nodeDict, edgeDict, aliases);
        numEdges += addAliasEdges(arg2, ind2, seenNps, writer, nodeDict, edgeDict, aliases);
      }

      if (!isKb && aliasesOnly) continue;
      List<String> relationEdges = getEmbeddedRelations(relation, embeddings);
      for (String r : relationEdges) {
        r = prefix + r;
        int rel = edgeDict.getIndex(r);
        writeEdgeIfUnseen(ind1, ind2, rel, seenTriples, writer);
        numEdges++;
      }
    }
    return numEdges;
  }

  private void writeEdgeIfUnseen(int arg1,
                                 int arg2,
                                 int rel,
                                 Set<IntTriple> seenTriples,
                                 FileWriter writer) throws IOException {
    if (seenTriples != null) {
      IntTriple triple = new IntTriple(arg1, arg2, rel);
      if (seenTriples.contains(triple)) return;
      seenTriples.add(triple);
    }
    writer.write(arg1 + "\t" + arg2 + "\t" + rel + "\n");
  }

  @VisibleForTesting
  protected int addAliasEdges(String np,
                              int np_index,
                              Set<String> seenNps,
                              FileWriter writer,
                              Dictionary nodeDict,
                              Dictionary edgeDict,
                              List<Pair<String, Map<String, List<String>>>> aliases)
  throws IOException {
    int numEdges = 0;
    if (seenNps.contains(np)) return numEdges;
    seenNps.add(np);
    for (Pair<String, Map<String, List<String>>> aliasSet : aliases) {
      String aliasRelation = aliasSet.getLeft();
      int aliasIndex = edgeDict.getIndex(aliasRelation);
      Map<String, List<String>> currentAliases = aliasSet.getRight();
      List<String> concepts = currentAliases.get(np);
      if (concepts == null) continue;
      for (String concept : concepts) {
        int concept_index = nodeDict.getIndex(concept);
        writer.write(np_index + "\t" + concept_index + "\t" + aliasIndex + "\n");
        numEdges++;
      }
    }
    return numEdges;
  }

  @VisibleForTesting
  protected List<String> getEmbeddedRelations(String relation, Map<String, List<String>> embeddings) {
    List<String> relationEdges = null;
    if (embeddings != null) {
      relationEdges = embeddings.get(relation);
      if (relationEdges != null && keepOriginalEdges) {
        List<String> relationEdgesCopy = Lists.newArrayList(relationEdges);
        relationEdgesCopy.add(relation);
        relationEdges = relationEdgesCopy;
      }
    }
    if (relationEdges == null) {
      relationEdges = Lists.newArrayList();
      relationEdges.add(relation);
    }
    return relationEdges;
  }

  private Map<String, List<String>> loadEmbeddings() throws IOException {
    Map<String, List<String>> embeddings = null;
    if (embeddingsFile != null) {
      System.out.println("Reading embeddings from file " + embeddingsFile);
      embeddings = fileUtil.readMapListFromTsvFile(embeddingsFile);
    }
    return embeddings;
  }

  public String getRelationFile() {
    return relationFile;
  }

  public String getRelationPrefix() {
    return relationPrefix;
  }

  public boolean getIsKb() {
    return isKb;
  }

  public String getAliasFile() {
    return aliasFile;
  }

  public boolean getAliasesOnly() {
    return aliasesOnly;
  }

  public String getAliasRelation() {
    return aliasRelation;
  }

  public String getEmbeddingsFile() {
    return embeddingsFile;
  }

  public boolean getKeepOriginalEdges() {
    return keepOriginalEdges;
  }

  private RelationSet(Builder builder) {
    relationFile = builder.relationFile;
    relationPrefix = builder.relationPrefix;
    isKb = builder.isKb;
    aliasFile = builder.aliasFile;
    aliasesOnly = builder.aliasesOnly;
    if (builder.aliasRelation == null) {
      aliasRelation = "@ALIAS@";
    } else {
      aliasRelation = builder.aliasRelation;
    }
    aliasFileFormat = builder.aliasFileFormat;
    embeddingsFile = builder.embeddingsFile;
    keepOriginalEdges = builder.keepOriginalEdges;
    fileUtil = builder.fileUtil;
  }

  public static RelationSet fromFile(String filename) throws IOException {
    return fromFile(filename, new FileUtil());
  }

  @VisibleForTesting
  protected static RelationSet fromFile(String filename, FileUtil fileUtil) throws IOException {
    return fromReader(fileUtil.getBufferedReader(filename), fileUtil);
  }

  public static RelationSet fromReader(BufferedReader reader) throws IOException {
    return fromReader(reader, new FileUtil());
  }

  @VisibleForTesting
  protected static RelationSet fromReader(BufferedReader reader, FileUtil fileUtil) throws IOException {
    Builder builder = new Builder();
    builder.setFileUtil(fileUtil);
    String line;
    while ((line = reader.readLine()) != null) {
      String[] fields = line.split("\t");
      String parameter = fields[0];
      String value = fields[1];
      if (parameter.equalsIgnoreCase("relation file")) {
        builder.setRelationFile(value);
      } else if (parameter.equalsIgnoreCase("relation prefix")) {
        builder.setRelationPrefix(value);
      } else if (parameter.equalsIgnoreCase("is kb")) {
        builder.setIsKb(Boolean.parseBoolean(value));
      } else if (parameter.equalsIgnoreCase("alias file")) {
        builder.setAliasFile(value);
      } else if (parameter.equalsIgnoreCase("aliases only")) {
        builder.setAliasesOnly(Boolean.parseBoolean(value));
      } else if (parameter.equalsIgnoreCase("alias relation")) {
        builder.setAliasRelation(value);
      } else if (parameter.equalsIgnoreCase("alias file format")) {
        builder.setAliasFileFormat(value);
      } else if (parameter.equalsIgnoreCase("embeddings file")) {
        builder.setEmbeddingsFile(value);
      } else if (parameter.equalsIgnoreCase("keep original edges")) {
        builder.setKeepOriginalEdges(Boolean.parseBoolean(value));
      }
    }
    return builder.build();
  }

  public static class Builder {
    private String relationFile;
    private String relationPrefix;
    private boolean isKb;
    private String aliasFile;
    private boolean aliasesOnly;
    private String aliasRelation;
    private String aliasFileFormat;
    private String embeddingsFile;
    private boolean keepOriginalEdges;
    private FileUtil fileUtil = new FileUtil();

    public Builder() {}
    public Builder setRelationFile(String r) {this.relationFile = r; return this;}
    public Builder setRelationPrefix(String p) {this.relationPrefix = p; return this;}
    public Builder setIsKb(boolean isKb) {this.isKb = isKb; return this;}
    public Builder setAliasFile(String aliasFile) {this.aliasFile = aliasFile; return this;}
    public Builder setAliasesOnly(boolean a) {this.aliasesOnly = a; return this;}
    public Builder setAliasRelation(String a) {this.aliasRelation = a; return this;}
    public Builder setAliasFileFormat(String a) {this.aliasFileFormat = a; return this;}
    public Builder setEmbeddingsFile(String e) {this.embeddingsFile = e; return this;}
    public Builder setKeepOriginalEdges(boolean k) {this.keepOriginalEdges = k; return this;}
    public Builder setFileUtil(FileUtil f) {this.fileUtil = f; return this;}

    public RelationSet build() {
      return new RelationSet(this);
    }
  }
}

File: src/main/scala/edu/cmu/ml/rtw/pra/graphs/RelationSet.scala
package edu.cmu.ml.rtw.pra.graphs

import edu.cmu.ml.rtw.pra.experiments.Outputter
import com.mattg.util.Dictionary
import com.mattg.util.FileUtil
import com.mattg.util.LineFilter
import com.mattg.util.IntTriple
import com.mattg.util.JsonHelper
import com.mattg.util.Pair

import java.io.BufferedReader
import java.io.DataOutputStream
import java.io.FileWriter

import scala.collection.mutable

import org.json4s._

class RelationSet(params: JValue, outputter: Outputter, fileUtil: FileUtil = new FileUtil) {
  implicit val formats = DefaultFormats

  // Fields dealing with the relations themselves.

  // The file containing the relation triples.
  val relationFile = (params \ "relation file").extract[String]

  // If not null, prepend this prefix to all relation strings in this set.
  val relationPrefix = JsonHelper.extractWithDefault(params, "relation prefix", null: String)

  // KB relations or surface relations?  The difference is only in the alias relation format (and
  // in the relation file format).
  val isKb = JsonHelper.extractWithDefault(params, "is kb", false)

  // Fields specific to KB relations.

  // If this is a KB relation set, this file contains the mapping from entities to noun phrases.
  val aliasFile = JsonHelper.extractWithDefault(params, "alias file", null: String)

  // What should we call the alias relation?  Defaults to "@ALIAS@".  Note that we specify this for
  // each set of _KB_ relations, not for each set of _surface_ relations.  If you want something
  // more complicated, like a different name for each (KB, surface) relation set pair, you'll have
  // to approximate it with relation prefixes.
  val aliasRelation = JsonHelper.extractWithDefault(params, "alias relation", "@ALIAS@")

  // Determines how we try to read the alias file.  We currently allow two values: "freebase" and
  // "nell".  The freebase format is a little complicated the nell format is just a list of
  // (concept, noun phrase) pairs that we read into a map.  The "nell" format is generic enough
  // that it could be used for more than just NELL KBs, probably.
  val aliasFileFormat = JsonHelper.extractWithDefault(params, "alias file format", null: String)


  // Fields specific to surface relations.

  // If this is true, we only add the alias edges that get generated from the set, not the
  // relations themselves (either from a KB relation set or a surface relation set).  This is still
  // somewhat experimental, but there are some cases where you might want to try this.
  val aliasesOnly = JsonHelper.extractWithDefault(params, "aliases only", false)

  // Fields for embeddings.

  // The file where we can find embeddings for the relations in this set.
  val embeddingsFile = JsonHelper.extractWithDefault(params, "embeddings file", null: String)

  // If we are embedding the edges, should we replace the original edges or just augment them?
  val keepOriginalEdges = JsonHelper.extractWithDefault(params, "keep original edges", false)

  val replaceRelationsWith = JsonHelper.extractWithDefault(params, "replace relations with", null: String)

  /**
   * Get the set of aliases specified by this KB relation set.
   */
  def getAliases(): Map[String, Seq[String]] = {
    if (aliasFile == null) {
      return Map()
    }
    aliasFileFormat match {
      case "freebase" => {
        fileUtil.readMapListFromTsvFile(aliasFile, 3, false, new LineFilter() {
          override def filter(fields: Array[String]): Boolean = {
            if (fields.length != 4) return true
            return false
          }
        })
      }
      case "nell" => {
        fileUtil.readInvertedMapListFromTsvFile(aliasFile, 1000000)
      }
      case other => throw new IllegalStateException("Unrecognized alias file format")
    }
  }

  def writeRelationEdgesToGraphFile(
    plainTextWriter: Option[FileWriter],
    binaryWriter: Option[DataOutputStream],
    seenTriples: mutable.HashSet[(Int, Int, Int)],
    prefixOverride: String,
    seenNps: mutable.HashSet[(String)],
    aliases: Seq[(String, Map[String, Seq[String]])],
    nodeDict: Dictionary,
    edgeDict: Dictionary
  ): Int = {
    outputter.info(s"Adding edges from relation file $relationFile")
    val prefix = {
      if (prefixOverride != null) {
        prefixOverride
      } else if (relationPrefix != null) {
        relationPrefix
      } else {
        ""
      }
    }
    var i = 0
    var numEdges = 0
    for (line <- fileUtil.getLineIterator(relationFile)) {
      i += 1
      fileUtil.logEvery(1000000, i)
      val fields = line.split("\t")
      // TODO(matt): Maybe make this relation file format a configurable field?
      val (arg1, arg2, relation) = if (isKb) {
        // KB relations are formated (S, O, V).
        (fields(0), fields(1), fields(2))
      } else {
        // And surface relations are formatted (S, V, O).
        (fields(0), fields(2), fields(1))
      }
      val ind1 = nodeDict.getIndex(arg1)
      val ind2 = nodeDict.getIndex(arg2)

      if (!isKb) {
        numEdges += addAliasEdges(arg1, ind1, seenNps, plainTextWriter, binaryWriter, nodeDict, edgeDict, aliases)
        numEdges += addAliasEdges(arg2, ind2, seenNps, plainTextWriter, binaryWriter, nodeDict, edgeDict, aliases)
      }

      if (!aliasesOnly) {
        val replaced = replaceRelation(relation)
        val relationEdges = getEmbeddedRelations(replaced, loadEmbeddings())
        for (relation <- relationEdges) {
          val prefixed_relation = prefix + relation
          val relation_index = edgeDict.getIndex(prefixed_relation)
          writeEdgeIfUnseen(ind1, ind2, relation_index, seenTriples, plainTextWriter, binaryWriter)
          numEdges += 1
        }
      }
    }
    numEdges
  }

  def writeEdgeIfUnseen(
      arg1: Int,
      arg2: Int,
      rel: Int,
      seenTriples: mutable.HashSet[(Int, Int, Int)],
      plainTextWriter: Option[FileWriter],
      binaryWriter: Option[DataOutputStream]) {
    if (seenTriples != null) {
      val triple = (arg1, arg2, rel)
      if (seenTriples.contains(triple)) return
      seenTriples.add(triple)
    }
    writeTriple(arg1, arg2, rel, plainTextWriter, binaryWriter)
  }

  def addAliasEdges(
      np: String,
      npIndex: Int,
      seenNps: mutable.HashSet[String],
      plainTextWriter: Option[FileWriter],
      binaryWriter: Option[DataOutputStream],
      nodeDict: Dictionary,
      edgeDict: Dictionary,
      aliases: Seq[(String, Map[String, Seq[String]])]): Int = {
    var numEdges = 0
    if (seenNps.contains(np)) return numEdges
    seenNps.add(np)
    for (aliasSet <- aliases) {
      val aliasRelation = aliasSet._1
      val aliasIndex = edgeDict.getIndex(aliasRelation)
      val currentAliases = aliasSet._2
      val concepts = currentAliases.getOrElse(np, Nil)
      for (concept <- concepts) {
        val conceptIndex = nodeDict.getIndex(concept)
        writeTriple(npIndex, conceptIndex, aliasIndex, plainTextWriter, binaryWriter)
        numEdges += 1
      }
    }
    return numEdges
  }

  def writeTriple(
    source: Int,
    target: Int,
    relation: Int,
    plainTextWriter: Option[FileWriter],
    binaryWriter: Option[DataOutputStream]
  ) {
    plainTextWriter.foreach(_.write(s"${source}\t${target}\t${relation}\n"))
    binaryWriter.foreach(w => {
      w.writeInt(source)
      w.writeInt(target)
      w.writeInt(relation)
    })
  }

  def replaceRelation(relation: String): String = {
    if (replaceRelationsWith != null) {
      replaceRelationsWith
    } else {
      relation
    }
  }

  def getEmbeddedRelations(relation: String, embeddings: Map[String, List[String]]) = {
    if (embeddings != null) {
      if (keepOriginalEdges) {
        relation :: embeddings.getOrElse(relation, Nil)
      } else {
        embeddings.getOrElse(relation, List(relation))
      }
    } else {
      List(relation)
    }
  }

  def loadEmbeddings(): Map[String, List[String]] = {
    if (embeddingsFile != null) {
      outputter.info("Reading embeddings from file " + embeddingsFile)
      fileUtil.readMapListFromTsvFile(embeddingsFile).mapValues(_.toList)
    } else {
      null
    }
  }
}
